File: apps/minecraft/shared/src/main/java/crabcraft/net/crabUtilities/vanish/VanishBridgeProtocol.java
package crabcraft.net.crabUtilities.vanish;

/** Wire format used to report EssentialsX vanish state from Paper to Velocity. */
public final class VanishBridgeProtocol {

    public static final String CHANNEL = "crabcraft:vanish";
    private static final byte VERSION = 1;
    private static final byte VISIBLE = 0;
    private static final byte VANISHED = 1;

    private VanishBridgeProtocol() {
    }

    public static byte[] status(boolean vanished) {
        return new byte[]{VERSION, vanished ? VANISHED : VISIBLE};
    }

    public static boolean decode(byte[] payload) {
        if (payload == null || payload.length != 2 || payload[0] != VERSION) {
            throw new IllegalArgumentException("Invalid vanish status payload");
        }
        return switch (payload[1]) {
            case VISIBLE -> false;
            case VANISHED -> true;
            default -> throw new IllegalArgumentException("Unknown vanish status value");
        };
    }
}

File: apps/minecraft/shared/src/main/kotlin/crabcraft/net/crabUtilities/vanish/VanishBridgeProtocol.kt
package crabcraft.net.crabUtilities.vanish

/** Wire format used to report EssentialsX vanish state from Paper to Velocity. */
object VanishBridgeProtocol {
    const val CHANNEL = "crabcraft:vanish"
    private const val VERSION: Byte = 1
    private const val VISIBLE: Byte = 0
    private const val VANISHED: Byte = 1

    @JvmStatic fun status(vanished: Boolean): ByteArray = byteArrayOf(VERSION, if (vanished) VANISHED else VISIBLE)

    @JvmStatic
    fun decode(payload: ByteArray?): Boolean {
        require(payload != null && payload.size == 2 && payload[0] == VERSION) {
            "Invalid vanish status payload"
        }
        return when (payload[1]) {
            VISIBLE -> false
            VANISHED -> true
            else -> throw IllegalArgumentException("Unknown vanish status value")
        }
    }
}
